c7n/resources/s3.py

```diff
@@ -2494,3 +2494,124 @@ def process_bucket(self, bucket):
+
+              - name: s3-enable-default-encryption-aes256
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+    """
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'type': {'enum': ['set-bucket-encryption']},
+            'key': {'type': 'string'}
+        }
+    }
+
+    permissions = ( 's3:PutEncryptionConfiguration',
+                    's3:GetEncryptionConfiguration')
+
+    def process(self, buckets):
+        with self.executor_factory(max_workers=3) as w:
+            futures = {}
+            results = []
+
+            config = { 'Rules': [
+                    {
+                        'ApplyServerSideEncryptionByDefault': {
+                            'SSEAlgorithm': 'AES256'
+                        }
+                    }
+                ]}
+
+            if 'key' in self.data:
+                config['Rules'][0]['ApplyServerSideEncryptionByDefault']['SSEAlgorithm'] = 'aws:kms'
+                config['Rules'][0]['ApplyServerSideEncryptionByDefault']['KMSMasterKeyID'] = self.data['key']
+
+            for b in buckets:
+                futures[w.submit(self.process_bucket, b, config)] = b
+
+            for future in as_completed(futures):
+                if future.exception():
+                    bucket = futures[future]
+                    self.log.error('error enabling bucket encryption: %s\n%s',
+                                   bucket['Name'], future.exception())
+                results += filter(None, [future.result()])
+
+            return results
+
+    def process_bucket(self, bucket, config):
+        s3 = bucket_client(local_session(self.manager.session_factory), bucket)
+        b = bucket['Name']
+        try:
+            r = s3.put_bucket_encryption(
+                    Bucket=b,
+                    ServerSideEncryptionConfiguration=config
+                )
+        except Exception as e:
+            raise
+
+@actions.register('delete-bucket-encryption')
+class DeleteBucketEncryption(BucketActionBase):
+    """Action disables default encryption on S3 buckets
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-dsable-default-encryption-kms
+                resource: s3
+                actions:
+                  - type: delete-bucket-encryption
+    """
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'type': {'enum': ['delete-bucket-encryption']}
+        }
+    }
+
+    permissions = ( 's3:PutEncryptionConfiguration',
+                    's3:GetEncryptionConfiguration')
+
+
+    def process(self, buckets):
+        with self.executor_factory(max_workers=3) as w:
+            futures = {}
+            results = []
+
+            for b in buckets:
+                futures[w.submit(self.process_bucket, b)] = b
```

Comment:
handle exceptions here instead of in process_bucket.  Exceptions can be thrown when getting a client or deleting the bucket, both will be caught here where as in the process_bucket function, client creation exceptions will bubble up